res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,890 @@
+        //       Mixxx. The deck load buttons, however, are both on MIDI channel 16, so I need
+        //       a better way to determine this.
+        off: Prime4.rgbCode.whiteDark,
+        on: Prime4.rgbCode.white,
+    });
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+Prime4.Deck.prototype = new components.Deck();
+
+Prime4.shift = false;
+Prime4.shiftState = function(channel, control, value) {
+    Prime4.shift = value === 0x7F;
+    if (Prime4.shift) {
+        midi.sendShortMsg(0x90 + channel, control, 0x02);
+        Prime4.leftDeck.shift();
+        Prime4.rightDeck.shift();
+        Prime4.effectBank[0].shift();
+        Prime4.effectBank[1].shift();
+    } else {
+        midi.sendShortMsg(0x90 + channel, control, 0x01);
+        Prime4.leftDeck.unshift();
+        Prime4.rightDeck.unshift();
+        Prime4.effectBank[0].unshift();
+        Prime4.effectBank[1].unshift();
+    }
+};
+
+//========== PERFORMANCE PADS ==========//
+
+// Access the appropriate mode-select pad without remembering MIDI values
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.PadSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    const theContainer = this;
+
+    // Create component containers for each pad mode
+    const modes = new components.ComponentContainer({
+        "hotcue": new Prime4.hotcueMode(deck, offset),
+        "loop": new Prime4.loopMode(deck, offset),
+        //TODO: "autoloop": new Prime4.autoloopMode(deck, offset),
+        "roll": new Prime4.rollMode(deck, offset),
+        "sampler": new Prime4.samplerMode(deck, offset),
+    });
+
+    modes.forEachComponent(c => c.disconnect());
+
+    const controlToPadMode = control => {
+        let mode;
+        switch (control) {
+        case Prime4.padMode.HOTCUE:
+            mode = modes.hotcue;
+            break;
+        case Prime4.padMode.LOOP:
+            mode = modes.loop;
+            /*
+            // TODO: Implement autoloop mode
+            if (Prime4.PadSection.currentMode === modes.loop) {
+                mode = modes.autoloop;
+            } else {
+                mode = modes.loop;
+            }
+            */
+            break;
+        case Prime4.padMode.ROLL:
+            mode = modes.roll;
+            break;
+        case Prime4.padMode.SLICER:
+            mode = modes.sampler;
+            break;
+        }
+
+        return mode;
+    };
+
+    this.offset = offset;
+
+    this.padModeSelectLeds = new components.Component({
+        trigger: function() {
+            modes.forEachComponent(mode => {
+                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
+            });
+        },
+    }, false);
```

Comment:
my mistake, though that shouldn't be the primary issue. Lets also include the debug output. 
```suggestion
                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
                console.log(mode, mode.ledControl, mode.colourOn, mode.colourOff);
            }, false);
        },
    });
```

Reply:
Not quite sure what the `false` parameter is for here, but these are my findings:

If `false` is kept on line 736, the `console.log` statement prints `js:debug [Controller] [object Object] undefined undefined undefined` as well as the big error log starting with `"Could not convert argument 1 at"...`.

Else, if `false` is placed on line 734, I can't find the error log or `console.log`, but still get no LED output on the mode selector pads.

Reply:
> Not quite sure what the false parameter is for here

It defines whether the forEachComponent method recurses down into the subcomponents. Since we only want to iterate over the containers, not over the actual padbuttons within, this is needed. For that reason, we get the message printed 32 times (because 4 padmodes * 8 pads = 32). 

> If false is kept on line 736, the console.log statement prints js:debug [Controller] [object Object] undefined undefined undefined as well as the big error log starting with "Could not convert argument 1 at"....

Yes, thats expected when iterating over the buttons since those obviously don't have the required properties (since they're undefined, but mixxx expects them to be a number, Qt warns that it doesn't know what to do with them). 

> Else, if false is placed on line 734, I can't find the error log or console.log, but still get no LED output on the mode selector pads.

Well thats at least progress. Can you comment out the `midi.sendShortMsg` call for now? It could be that the JS engine just silently breaks when its being passed invalid parameters, which is why there is no log output in that case? Maybe just deliberately throwing in there is also an option. I have the suspicion that `forEachComponent` is not being executed on the componentcontainer like I expected it too. In that case, you can try something like this:

```js
        trigger: function() {
            for mode in Object.entries(modes) {
                console.log(mode, mode.ledControl, mode.colourOn, mode.colourOff);
                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
            }
        },
```